Source/Core/DiscIO/VolumeVerifier.cpp

```diff
@@ -0,0 +1,648 @@
+    CheckCorrectlySigned(PARTITION_NONE, GetStringT("This title is not correctly signed."));
+  CheckDiscSize();
+  CheckMisc();
+}
+
+std::vector<Partition> VolumeVerifier::CheckPartitions()
+{
+  const std::vector<Partition> partitions = m_volume.GetPartitions();
+  if (partitions.empty())
+  {
+    if (m_volume.GetVolumeType() != Platform::WiiWAD &&
+        !m_volume.GetFileSystem(m_volume.GetGamePartition()))
+    {
+      AddProblem(Severity::High, GetStringT("The filesystem is invalid or could not be read."));
+      return {};
+    }
+    return {m_volume.GetGamePartition()};
+  }
+
+  std::optional<u32> partitions_in_first_table = m_volume.ReadSwapped<u32>(0x40000, PARTITION_NONE);
+  if (partitions_in_first_table && *partitions_in_first_table > 8)
+  {
+    // Not sure if 8 actually is the limit, but there certainly aren't any discs
+    // released that have as many partitions as 8 in the first partition table.
+    // The only game that has that many partitions in total is Super Smash Bros. Brawl,
+    // and that game places all partitions other than UPDATE and DATA in the second table.
+    AddProblem(Severity::Low,
+               GetStringT("There are too many partitions in the first partition table."));
+  }
+
+  std::vector<u32> types;
+  for (const Partition& partition : partitions)
+  {
+    const std::optional<u32> type = m_volume.GetPartitionType(partition);
+    if (type)
+      types.emplace_back(*type);
+  }
+
+  if (std::find(types.cbegin(), types.cend(), PARTITION_UPDATE) == types.cend())
+    AddProblem(Severity::Low, GetStringT("The update partition is missing."));
+
+  if (std::find(types.cbegin(), types.cend(), PARTITION_DATA) == types.cend())
+    AddProblem(Severity::High, GetStringT("The data partition is missing."));
+
+  const bool has_channel_partition =
+      std::find(types.cbegin(), types.cend(), PARTITION_CHANNEL) != types.cend();
+  if (ShouldHaveChannelPartition() && has_channel_partition)
+    AddProblem(Severity::Medium, GetStringT("The channel partition is missing."));
+
+  if (ShouldHaveMasterpiecePartitions() &&
+      types.cend() ==
+          std::find_if(types.cbegin(), types.cend(), [](u32 type) { return type >= 0xFF; }))
+  {
+    // i18n: This string is referring to a game mode in Super Smash Bros. Brawl called Masterpieces
+    // where you play demos of NES/SNES/N64 games. Official translations:
+    // 名作トライアル (Japanese), Masterpieces (English), Meisterstücke (German), Chefs-d'œuvre
+    // (French), Clásicos (Spanish), Capolavori (Italian), 클래식 게임 체험판 (Korean).
+    // If your language is not one of the languages above, consider leaving the string untranslated
+    // so that people will recognize it as the name of the game mode.
+    AddProblem(Severity::Medium, GetStringT("The Masterpiece partitions are missing."));
+  }
+
+  for (const Partition& partition : partitions)
+  {
+    if (m_volume.GetPartitionType(partition) == std::optional<u32>(PARTITION_UPDATE) &&
```

Comment:
Does `std::optional` not have a comparison operator against the contained type? Seems weird you'd have to type it out like that (C# Nullables just compare `false` if they don't hold any value)

Reply:
There is actually a comparison operator like that, I just didn't know about it. I'll change it.